### DISCUSSION TITLE
Merging scripts.lst

### DISCUSSION THREAD

**Discussion by Lexx2k** [sfall-team/sfall#352] [Ideas]
I just realized, this is probably something that is better suited for the new "discussions" than the issue tickets...
_______________________________________

Random idea for a bit better mod compatibility: (if gl scripts can't be used)  Can't we do something like a new lookup table that basically says "script name ModXScpt.int is line 50 in scripts.lst"?

Example of Use:
We are adding a placeholder script entry to MainMod's script.lst file
`test.int ; Reserved for ModX` - it is on line 50.
MainMod could have 100 or more of such placeholders as reserved for other mods.

Then in Fallout2\mods\ModX\ folder we have a new file called ModScripts.ini that says ...
`Script01=50|ModXScpt.int`
... which is the supposed line number of the script, and the script name.

Then, when booting up the game with MainMod and ModX together, the `ModScripts.ini` file will be read and automatically assigns ModXScpt.int file to the line number 50 of MainMod.

Is a bit complicated, but would enhance compatibility, imo. Just a random thought I had when thinking about overriding scrname.msg ... The NPC / Object / Whatever with the new script attached could then be spawned via global script and we wouldn't even have to touch map files.

**Comment by FakelsHub** [comment]
Override line.
`Script01=name.int, localvar, scrname`

**Comment by FakelsHub** [comment]
In general, the concept of the engine is not designed for multi-mods.

I'd rather concentrate my efforts on making the engine more modifiable for a single global mod than introducing support for compatible connection of different mods.

**Reply by Lexx2k** [reply]
It's just a random idea I had this morning, not something I totally need right now.

I think it would be useful for smaller mods which don't try to overhaul half of the game and aren't worth the effort of releasing a new repack. The "main mod author" (for UPU / RPU as example) could just provide a few placeholders which then can be "hooked into" by other mods. No need to directly integrate them (most mod authors wouldn't do this anyways, let's be honest).

**Reply by phobos2077** [reply]
Back when I was still modding FO2 I had exactly the same idea about modularity. Yes, the original engine is not suited for multi-mods, but sfall IS. They way hook scripts and global scripts are made - they support multiple smaller mods working together. So I wouldn't simply brush this discussion aside as "non important".

If you want to focus on one large mod - that's fine, but don't assume everyone has the same goal.

**Reply by burner1024** [reply]
I'm all pro multi mod, but I don't quite understand the idea yet.

**Reply by Lexx2k** [reply]
Main mod has placeholder scripts, protos, etc that can be used (overwritten) by other mods. It's really that simple. You can already do it now, but it's not very practical, because then the sub-mod _still_ has to constantly update certain files if the main mod changes something _and_ it is pretty much impossible to keep compatibility with other smaller mods.

See it as odd hooks or something that can be used by other mods.

With my above stated idea, it would - for example - be possible to add new NPC scripts and dialogues into the main mod without having to actually touch the main mod. You just use one of the placeholder scripts and variables.

I'm not saying this is the best idea ever, but it could be suitable for smaller things, which then - if the mod makers don't all use the same placeholders - would remain compatible to each other.

However, considering the Fo2 modding community is pretty much just us at this point, it's also a matter of "is it really worth it" -- I can't answer that question.

**Comment by burner1024** [comment]
What exactly bothers you, adding entries to scripts.lst? scrname.msg? Creating a bunch of scrips that include a common template?

> gl scripts can't be used

when is that?

**Reply by Lexx2k** [reply]
Sorry, I don't know how to explain it any better without repeating myself the third time.


> when is that?

You can't add a new NPC with _new_ dialog via a global script. At least not as far as I am aware.

**Reply by Lexx2k** [reply]
Let's take the UPU as example. If I would make a quest mod for that, it is no problem. I have to keep it updated with the UPU updates but that's it.

Now if someone else comes along and also wants to make a quest mod for the UPU that is, again, no problem... UNLESS a player wants to install _both_ quest mods. Now that won't work anymore unless someone merges them.

This issue can be circumvented if the UPU by default adds placeholders which quest mod #1 and quest mod #2 can use. As long as both quest mods don't use the same placeholders, they will be compatible with each other.


This is the basic idea for the placeholders which would be reserved for and overwritten by these sub-mods.

Of course this is highly theoretical and right now there is nobody who is going to do mods like that. Which is why I've just added this idea to discussions, and so it won't be forgotten in the future.

**Reply by burner1024** [reply]
OK, now I understand.
Well, of course there nobody who is going it do mods like that, if it's impossible in the first place.
However, I don't like messing with inis and as proposed this requires coordination between authors of the main mod, and the smaller mods, all of them together.
(In fact, I don't like scripts.lst at all. I wish scripts would just load messages by name.)
But, there's precedent of [add_extra_msg_file](https://sfall.bgforge.net/sfall-funcx-macros/#add_extra_msg_file), which allows to register custom messages on the fly. Maybe the same would be possible for scripts? It wouldn't require placeholders.

**Reply by FakelsHub** [reply]
> But, there's precedent of add_extra_msg_file, which allows to register custom messages on the fly. Maybe the same would be possible for scripts? It wouldn't require placeholders.

just a sketch of an example implementation of how this will work in the game.
```
reg_index = reg_script(rg_<name>.int, localvar, srcname, bindToObj, name.msg)
set_obj_local_var(bindToObj, var_index, reg_index)

reg_index - the number under which the script was registered to access its msg file for vanilla dialog functions.

in register script:
msg_index = get_local_var(var_index)
gsay_message(msg_index, msg_num, reaction)
```

**Reply by FakelsHub** [reply]
yes

**Reply by burner1024** [reply]
Isn't current `gl_` with `game_loaded` enough?

**Reply by FakelsHub** [reply]
это не глобальный скрипт, чтобы он имел "gl" префикс, и такие вещи желательно дифференцировать.
_this is not a global script, so that it has a "gl" prefix, and such things are desirable to differentiate._

**Reply by burner1024** [reply]
I'm not saying the new script should start wit `gl_`. On the contrary, I'm saying I don't see a reason to require any prefix such as `rg_`. Care to elaborate?

**Reply by FakelsHub** [reply]
> Care to elaborate?

такие вещи желательно дифференцировать.

**Reply by burner1024** [reply]
Что от чего дифференцировать-то?

**Comment by phobos2077** [comment]
This idea sounds very useful. I remember years ago thinking about different solutions to this a lot. Will need some hackery though, as the concept of LST files is so deeply integrated with the engine.

**Comment by phobos2077** [comment]
See my idea from this issue, it covers scripts as well as all other assets that use LST indexing: https://github.com/sfall-team/sfall/issues/289#issuecomment-1577371316
